kjdoc/tour.md

```diff
@@ -975,6 +975,114 @@ As of this writing, KJ supports C++20 coroutines and Coroutines TS coroutines, t
 
 KJ prefers C++20 coroutines when both implementations are available.
 
+### Asynchronous environments
+
+Environments are ambiently available storage, of a type specified by the user, which "follow" user callbacks scheduled via functions/syntax like evalLater(), promise.then(), startFiber(), co_await, and more.
+
+Environments do not cross threads. Type safety is as safe as kj::downcast<T>() makes it -- in debug mode, you'll probably get an assertion if you set an environment of type T and try to get it later as type U -- but in release mode you're on your own.
```

Comment:
It's not clear here where "type safety" is involved.

Does this mean there's only one active environment at any particular time, it has one specific type, and you have to use the correct type when calling `getEnvironment()`?

I would argue it's worth making this throw in release mode too.

But meanwhile I kind of expected that the "environment" actually has a slot for every possible type. `runInEnvironment` would only replace one of these slots, leaving the rest with their existing value. This is probably what we want since there may be multiple use cases for environments within the same codebase and forcing them all to share a single environment type might be cumbersome.

Reply:
Hmm, yes, I think this paragraph must have survived from a much earlier implementation, and I didn't revise it.

The current implementation has a slot for every possible type, using a `HashMap<void*, Own<Environment>>`. The `void*` is a "type key" implemented as the address of a static variable inside a function template parameterized by the user's concrete type, T. `Environment` is a base class whose derived implementations are holders of T. Since `runInEnvironment()` and `getEnvironment()` both know what T is, they each calculate the same type key, and thus the downcast should be safe. If we pass a type U to `getEnvironment()`, the HashMap will just tell us no value exists for that type key, so my comment here is wrong.